### PR TITLE
[Alarms & Timers] Support for Monday as first day of week

### LINKED ISSUE #1780: Alarm: Day of week settings should not always start on sunday
Right now the day of week settings in the alarm app always has Sunday as first day of the week. Dependent on the locale it should switch to Monday as first day. Problem is: This seems to be not yet provided by the locale or am I missing something?

So the alternatives are:
1. Add week-starts-on-monday to locale
2. Add week-starts-on-monday to alarm settings
Any opinions on this?

### PR DESCRIPTION
**Pull request by alessandrococco**
_PR updated, see comments below_

~~This PR improves the UI of the default alarms app.~~

~~- Show localized abbreviations for days of week (current version always uses english abbreviations)~~
~~- An alarm on repeat Monday to Friday is labeled "repeat on workdays"~~
~~- An alarm on repeat Saturday and Sunday is labeled "repeat on weekends"~~

~~Some screenshots:~~

image 

image 

Any feedback/suggestions?

---

Closes #1780

### CODE REVIEW

**Comment by storm64** [comment]
This looks nice, I really like to have the dow displayed on the selection screen to decide between the alarms. I also like how you implemented workdays and weekends.

But I'm not 100% sure about the change from the 🔁-symbol to a `Repeat` as text element:
1. On some languages this might get very long, e.g. German: `Repeat` (6) to `Wiederholen` (11) / `Wiederholungen` (14)
2. And I don't see the necessity of a second line with `Once` for a non repeating alarm.

Like to hear your thoughts on this.

**Comment by gfwilliams** [comment]
Yes, I don't like this either I'm afraid. Moving from what was usually a single line of big text to 2 small lines pretty much all the time seems like a big step backwards. There are a bunch of users that have been pretty passionate about trying to get the font in the menus much bigger, and I can pretty much guarantee they won't like this.

I don't really see the problem with the icon. Maybe I'm wrong but it feels like a bunch of recent changes haven't really been discussed with anyone to see if they actually want them? It's just changing stuff to what you personally prefer.

I'm just conscious that when stuff breaks or changes, it'll just be me that has to deal with the complaints.

**Comment by alessandrococco** [comment]
Thank you @gfwilliams for your feedback. Sorry if I bothered you, I should limit my enthusiasm 😄 

> Moving from what was usually a single line of big text to 2 small lines pretty much all the time seems like a big step backwards. There are a bunch of users that have been pretty passionate about trying to get the font in the menus much bigger, and I can pretty much guarantee they won't like this.

In a discussion in [my fork](https://github.com/alessandrococco/BangleApps/issues/5#issuecomment-1107467519) I proposed an option for optionally showing the dow in the alarm list (imo this is an important feature: the user can see at glance when an alarm is on without the need to open it) but didn't add to the PR yet (I just noticed I marked _this_ PR "ready for review" by mistake)

> I don't really see the problem with the icon. 

I was having trouble getting the icon work right on two lines, this is my fault.

> Maybe I'm wrong but it feels like a bunch of recent changes haven't really been discussed with anyone to see if they actually want them

What is the best way to proceed? Do you prefer a **new issue** here or a **new forum thread**?

-----

Anyway, I'll fix my PR:

- I'll remove the dow in the alarm list but I'll keep the localized dow in the edit alarm menu
- I'll add the support for "Start Week On Mon/Sun" ( #1780, I'll cherry-pick some commits from my other branch) 

About the dow in the alarm list, updated _repeat_ label, etc. I'll open a issue/thread to discuss about it :-)


----

English is obviously not my mother tongue but I hope you understand what I'm trying to say 😃

**Comment by gfwilliams** [comment]
> Sorry if I bothered you, I should limit my enthusiasm

Not at all! It's great to have all these contributions - but I tend to get a lot of complaints from users when things change, so it's good to make sure that when they change, as many people as possible think it's an improvement :)

> I proposed an option for optionally showing the dow

I think showing DOW is fine for now (as long as it's not shown when *all* days are selected). I do like the idea of 'weekends/weekdays' too but if it takes up *more* space than just `SMT_TFS` it may not actually be making things easier to read, especially as when they all say `SMT_TFS` it's easy to compare.

Maybe removing the checkbox on the right (maybe replacing it with an icon that's alarm/timer/disabled-alarm/disabled-timer) would give us more space to get everything on a line?

> What is the best way to proceed? Do you prefer a new issue here or a new forum thread?

Ideally an issue would be the right place, **but** most people (and the ones that will complain) seem to sit mainly on the forum so I think it's best to ask there.

> I was having trouble getting the icon work right on two lines

Ahh that's interesting. Could be a rendering problem, or maybe the icon just needs to be a few pixels smaller?

> PR changes

That'd be perfect, thanks!

> English is obviously not my mother tongue but I hope you understand what I'm trying to say

Absolutely. And wow - I couldn't have guessed based on your posts at all :)

**Comment by gfwilliams** [comment]
> Sorry if I bothered you, I should limit my enthusiasm

Not at all! It's great to have all these contributions - but I tend to get a lot of complaints from users when things change, so it's good to make sure that when they change, as many people as possible think it's an improvement :)

> I proposed an option for optionally showing the dow

I think showing DOW is fine for now (as long as it's not shown when *all* days are selected). I do like the idea of 'weekends/weekdays' too but if it takes up *more* space than just `SMT_TFS` it may not actually be making things easier to read, especially as when they all say `SMT_TFS` it's easy to compare.

Maybe removing the checkbox on the right (maybe replacing it with an icon that's alarm/timer/disabled-alarm/disabled-timer) would give us more space to get everything on a line?

> What is the best way to proceed? Do you prefer a new issue here or a new forum thread?

Ideally an issue would be the right place, **but** most people (and the ones that will complain) seem to sit mainly on the forum so I think it's best to ask there.

> I was having trouble getting the icon work right on two lines

Ahh that's interesting. Could be a rendering problem, or maybe the icon just needs to be a few pixels smaller?

> PR changes

That'd be perfect, thanks!

> English is obviously not my mother tongue but I hope you understand what I'm trying to say

Absolutely. And wow - I couldn't have guessed based on your posts at all :)

**Comment by alessandrococco** [comment]
PR updated! It includes:

- updated `date_utils` (with documentation!)
- localized dow in the alarm menu
- support for Monday as first day of week

I did some tests on my Bangle and it seems ok to me.

I also cherry-picked some commits from my other branches to include some minor _qol_ improvements, eg. a positive time zone will be shown as "+X" instead of "X". You can read about them in the commit messages, I kept all changes in different commits.

image

---

>  Could be a rendering problem, or maybe the icon just needs to be a few pixels smaller?

The icon did not align properly so I put it aside for a moment.. and forget about it 😅 

> Ideally an issue would be the right place, but most people (and the ones that will complain) seem to sit mainly on the forum so I think it's best to ask there.

Ok!

File: apps/alarm/app.js [hunk 1]
```diff
@@ -69,25 +78,27 @@ function showMainMenu() {
...
     '': { 'title': /*LANG*/'Days of Week' },
-    /*LANG*/'< Back' : () => onchange(dow)
+    /*LANG*/'< Back': () => onchange(dow)
   };
-  for (let i = 0; i < 7; i++) (i => {
-    let dayOfWeek = require("locale").dow({ getDay: () => i });
-    menu[dayOfWeek] = {
-      value: !!(dow&(1<<i)),
+  var dows = require("date_utils").dows(firstDayOfWeek);
+  for (var i = 0; i < 7; i++) (i => {
```
Comment (gfwilliams) [inline apps/alarm/app.js:84]: I'm merging anyway, but IMO `require("date_utils").dows(firstDayOfWeek).forEach(...)` is a bit tidier than `for (var i = 0; i < 7; i++) (i => { ... })(i)` now we have an array to work from

**Comment by gfwilliams** [comment]
Looks great - thanks! I'll update the apps/setting/README.md at this end

**Pull request merged**